### ISSUE TITLE
[slop~] 1st arg and 2nd inlet unclear to me

### ISSUE THREAD

**Issue by porres** [pure-data/pddp#242]
The [slop~] help file is not very descriptive on what the parameters are. We rely on the examples to get an idea, but no example really explores os explains the 1st argument or 2nd inlet in depth, and I just don't get what it does.

Help file could be improved in explaining that

**Comment by porres** [comment]
in the ADSR example I added to the help file I have [slop~ 1e+09] but this 1st argument does not seem to be necessary at all...

**Comment by porres** [comment]
So, if you have a huge 'slew' limit, then [slop~] acts as a regular lowpass filter, and it can be implemented in [fexpr~] as

**[fexpr~ $y + inc * ($x - $y)]**

where 'inc' is the freq input in radians per sample

I get this, what I don;t get is how it interplays with the other parameters when you have a slew limit of 0 or low...

**Comment by porres** [comment]
The slew limit examples need to have a big number for the 1st argument so it works properly, and I'm not sure why or how to explain it. If the number is low it interferes and slows down the slew.

**Comment by porres** [comment]
The slew limit examples (using 3rd and 5th inlets - or 2nd and 4th arg) need to have a big number for the 1st argument or 2nd inlet so it works properly, and I'm not sure why or how to explain it. If the number is low it interferes and slows down the slew. ALSO... if the 4th/6th inlet input is high, it ruins the slew limiting. So the 3rd/4th inlets are particularly sensitive to the other inlet's input and is ALWAYS Influenced by them!

And IF the slew limit value is HUGE, then the 1st arg / 2nd inlet makes [slop~] act as a regular lowpass filter, and it can be implemented in [fexpr~] as

[fexpr~ $y + inc * ($x - $y)]

where 'inc' is the freq input in radians per sample

I am not sure yet how this interplays with the 4th and 6th inlets. It seems they supersede. Like, for the 4th and 6th inlet or 3rd and 5th argument, the the 1st argument or 2nd inlet can be zero! And it seems it doesn't matter if it is a high number (tested with the ADSR example I added)

**Comment by porres** [comment]
In short. 

- 1st arg depends on slew limits (2nd/4th args) being huge to make [slop~] a lowpass filter.
- 1st arg is ignored and gets superseded by 3rd/5th args
- 2nd/4th args as slew limits depend on 1st arg being huge and  3rd/5th args being ZERO

**Comment by porres** [comment]
Looking at the source code, we have 

*sigin (1st inlet input signal);
*freqin (2nd inlet / cutoff: aka "inc" in rad/s);
*neglimit (3rd inlet / downward slew limit: aka "mindiff");
*negfreqinc (4th inlet / downward cutoff: aka "neginc" in rad/s);
*poslimit (5th inlet / upwards slew limit: aka "maxdiff");
*posfreqin (6th inlet / upwards cutoff: "posinc" in rad/s);

diff = sigin - last; // input minus last output

diffinc // I can't make sense of this name

if (diff > maxdiff); 
// if diff > upwards slew limit, therefore, if we exceed the slew limit
    diffinc = posinc * (diff - maxdiff) + inc * maxdiff;
// upwards cutoff * (diff - upslew) + cutoff * upslew


(I am not done)


if (diff > maxdiff);
     diffinc = posinc * (diff- maxdiff) + inc * maxdiff;
else if (diff < -mindiff);
     diffinc = neginc * (diff + mindiff) - inc * mindiff;
else diffinc = inc * diff;

**Comment by millerpuckette** [comment]
Hi Alexandre -

did you look at slop-tilde.htm in 8.topics?  That tries to explain it :)

I found that there wasn't a way to explain it fully in the help window - 
needed equation typesetting - and failed through to markdown :)

M

**Comment by porres** [comment]
> did you look at slop-tilde.htm in 8.topics? 

yup, but it was also a bit over my head, I'm having better luck figuring it out by checking the code and writing equivalent code in [fexpr~]. I'm making progress, but still have questions

**Comment by porres** [comment]
@millerpuckette , I edited my comment above and I guess changes haven't been sent to your email as a notice. So here's a new comment

> Now, there are combinations of parameters that are still missing. I wonder if there is any kind of combination that could still be meaningful and useful.

What can you tell me about this? 

I realize that is not just because you can set the filter to behave differently, that it might be useful.

Anyway, testing and checking the formula made the examples clear to me. I will now read again the 8.topics section to see if I find anything new.

**Comment by porres** [comment]
Hey, now I get tha [lop~] is pretty much completely equivalent to [slop~] for the same frequency input, wow!

**Comment by porres** [comment]
> I will now read again the 8.topics section to see if I find anything new.

I have to say that this HTML file doesn't help much as it doesn't really explain much about the formulas and how the parameters interact. I just can't see it.

> I found that there wasn't a way to explain it fully in the help window -
needed equation typesetting - and failed through to markdown :)

But the HTML reference is not quite rich in detail on what the filter equations are with these parameters and all (or maybe it is just not really clear to me). I get that maybe the help file is not the best place for it and if we have an HTML file then we can maybe explore it over there in details with math and all. But even without the math, it'd be good to make it clear on the help file how the filter works, with information I've been laying out here that is not really clear anywhere for me unless I read the source code, like the A) to D) listed items.
